src/algorithms/synthesis/encoding.cpp

```diff
@@ -0,0 +1,126 @@
+        }
+
+        // if the truth table function is already reversible, no encoding is necessary
+        if (outputFreq.size() == tt.cubeMap.size()) {
+            return;
+        }
+
+        // create a priority queue for building the Huffman tree
+        auto comp = [](const std::shared_ptr<MinHeapNode>& a, const std::shared_ptr<MinHeapNode>& b) {
+            return *a > *b;
+        };
+        std::priority_queue<std::shared_ptr<MinHeapNode>,
+                            std::vector<std::shared_ptr<MinHeapNode>>,
+                            decltype(comp)>
+                minHeap(comp);
+
+        // initialize the leaves of the Huffman tree from the output frequencies
+        for (const auto& [output, freq]: outputFreq) {
+            const auto requiredGarbage = static_cast<std::size_t>(std::ceil(std::log2(freq)));
+            minHeap.emplace(std::make_shared<MinHeapNode>(output, requiredGarbage));
+        }
+
+        // combine the nodes with the smallest weights until there is only one node left
+        while (minHeap.size() > 1U) {
+            // pop the two nodes with the smallest weights
+            const auto left = minHeap.top();
+            minHeap.pop();
+            const auto right = minHeap.top();
+            minHeap.pop();
+            // compute appropriate frequency to cover both nodes
+            const auto freq = std::max(left->freq, right->freq) + 1U;
+            // create new parent node
+            auto top   = std::make_shared<MinHeapNode>(TruthTable::Cube{}, freq);
+            top->left  = left;
+            top->right = right;
+            // add node to queue
+            minHeap.emplace(std::move(top));
+        }
+
+        const auto requiredGarbage = minHeap.top()->freq;
+        // determine encoding from Huffman tree
+        TruthTable::CubeMap encoding{};
+        minHeap.top()->traverse({}, encoding);
+
+        // resize all outputs to the correct size (by adding don't care values)
+        for (auto& [input, output]: encoding) {
+            output.resize(requiredGarbage);
+        }
+
+        // encode all the outputs
+        for (auto& [input, output]: tt.cubeMap) {
+            output = encoding[output];
+        }
+    }
+
+    auto augmentWithConstants(TruthTable& tt) -> void {
+        for (auto const& [input, output]: tt.cubeMap) {
+            const auto inputSize  = input.size();
+            const auto outputSize = output.size();
+            if (inputSize >= outputSize) {
+                continue;
+            }
+
+            const auto requiredConstants = outputSize - inputSize;
+            auto       newCube           = input;
+            newCube.reserve(outputSize);
+            for (std::size_t i = 0; i < requiredConstants; i++) {
+                newCube.insertZero();
+            }
+            auto nh  = tt.cubeMap.extract(input);
+            nh.key() = newCube;
+            tt.cubeMap.insert(std::move(nh));
+        }
+    }
+
+} // namespace syrec
```

Comment:
there is a newline missing at the end of this file.

Reply:
done 👍🏼